cv/lane_detection/src/lane_detection_inferencer.cpp

```diff
@@ -0,0 +1,242 @@
+      cuda_options_{OrtCUDAProviderOptions()},
+      session_options_{getOrtSessionOptions(1, cuda_options_)},
+      session_{env_, model_path_.c_str(), session_options_},
+      allocator_{Ort::AllocatorWithDefaultOptions()},
+      input_names_{session_.GetInputName(0, allocator_)},
+      output_names_{session_.GetOutputName(0, allocator_)},
+      input_dims_{
+          session_.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()},
+      output_dims_{
+          session_.GetOutputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()},
+      input_tensor_size_(vectorProduct(input_dims_)),
+      output_tensor_size_(vectorProduct(output_dims_)),
+      output_tensor_values_{std::vector<float>(output_tensor_size_)},
+      memory_info_{Ort::MemoryInfo::CreateCpu(
+          OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault)}
+{
+}
+
+cv::Mat LaneDetectionInferencer::getLaneDetectionMask(cv::Mat input_image)
+{
+    return postprocessImage(performInference(preprocessImage(input_image)));
+}
+
+cv::Mat LaneDetectionInferencer::preprocessImage(cv::Mat input_img)
+{
+    cv::Mat resized_img;
+    cv::resize(input_img, resized_img,
+               cv::Size(NETWORK_INPUT_WIDTH, NETWORK_INPUT_HEIGHT),
+               cv::INTER_LINEAR);
+
+    cv::Mat input_gray_img;
+    cv::cvtColor(resized_img, input_gray_img, cv::COLOR_BGR2GRAY);
+    cv::Mat edges = getEdgeChannel(input_gray_img);
+    cv::Mat edges_inv = getInverseEdgeChannel(edges);
+
+    cv::Mat normalized_edges;
+    cv::Mat normalized_edges_inv;
+    edges.convertTo(normalized_edges, CV_32F, 1.0f / 255.0f, 0.0f);
+    edges_inv.convertTo(normalized_edges_inv, CV_32F, 1.0f / 255.0f, 0.0f);
+
+    cv::Mat normalized_img;
+    resized_img.convertTo(normalized_img, CV_32F, 1.0f / 255.0f, 0.0f);
+
+    return convertHWC2CHW(
+        combineImages(normalized_img, normalized_edges, normalized_edges_inv));
+}
+
+cv::Mat LaneDetectionInferencer::performInference(cv::Mat preprocessed_img)
+{
+    std::vector<float> input_tensor_values(input_tensor_size_);
+    input_tensor_values.assign(preprocessed_img.begin<float>(),
+                               preprocessed_img.end<float>());
+
+    std::vector<Ort::Value> input_tensors;
+    std::vector<Ort::Value> output_tensors;
+
+    input_tensors.push_back(Ort::Value::CreateTensor<float>(
+        memory_info_, input_tensor_values.data(), input_tensor_size_,
+        input_dims_.data(), input_dims_.size()));
+
+    output_tensors.push_back(Ort::Value::CreateTensor<float>(
+        memory_info_, output_tensor_values_.data(), output_tensor_size_,
+        output_dims_.data(), output_dims_.size()));
+
+    session_.Run(Ort::RunOptions{nullptr}, input_names_.data(),
+                 input_tensors.data(), 1, output_names_.data(),
+                 output_tensors.data(), 1);
+
+    // TODO: do not use C style cast, look for safer way to do this
+    return cv::Mat(NETWORK_OUTPUT_HEIGHT, NETWORK_OUTPUT_WIDTH, CV_32FC1,
```

Comment:
@spencerteetaert 

Do you know what the underlying data is? I would prefer avoiding a cast if possible but I'm fine with it since everything works I guess

Reply:
As far as I can tell it's some Ort value class. Not sure what the actual datatype is, I would assume in this case it's a wrapper for array of float32 by default

Reply:
I'm fine with the cast then, we'll worry about fixing it later if it's problematic